### PR TITLE
Environment variables; access and parsing

### PR DESCRIPTION
**Pull request by lancelet**
Accessing and parsing environment variables

### CODE REVIEW

File: src/Data/Hocker/Process.hs [hunk 1]
```diff
@@ -46,6 +67,32 @@ newtype ProcessErrorDetector e m = ProcessErrorDetector
...
+    where
+      f _ _ _ = return ()
+  mappend a b = ProcessErrorDetector f
+    where
+      f :: MonadError e m => ExitCode -> StdOut -> StdErr -> m ()
+      f exitCode out err =
+        do
+          (unProcessErrorDetector a) exitCode out err
+          (unProcessErrorDetector b) exitCode out err
+          return ()
```
Comment (lancelet) [inline src/Data/Hocker/Process.hs]: Monoid instance. Following Gonzalez' advice to make things moar algebraic.

File: src/Data/Hocker/Process.hs [hunk 2]
```diff
@@ -100,9 +147,152 @@ runProcess cmd args stdin maybeEnv errDetectors =
...
+-- | Represents a set of environment variables.
+newtype EnvironmentVariables = EnvironmentVariables
+  { unEnvironmentVariables :: HashMap.HashMap Text Text }
+  deriving (Show, Eq)
+
+
+instance Monoid EnvironmentVariables where
+  mempty = EnvironmentVariables mempty
+  mappend a b =
+    EnvironmentVariables (unEnvironmentVariables a <> unEnvironmentVariables b)
```
Comment (luke-clifton) [inline src/Data/Hocker/Process.hs]: This is where `GeneralizedNewtypeDeriving` can be nice.

File: src/Data/Hocker/Process.hs [hunk 3]
```diff
@@ -100,9 +147,152 @@ runProcess cmd args stdin maybeEnv errDetectors =
...
+-- | Retrieves the current process environment variables.
+getEnvironment :: (MonadIO m) => m EnvironmentVariables
+getEnvironment =
+  do
+    sysenv <- liftIO SysEnv.getEnvironment
+    let
+      env' = ( EnvironmentVariables
+             . HashMap.fromList
+             . fmap ( \(a,b) -> (T.pack a, T.pack b) )
+             ) sysenv
```
Comment (luke-clifton) [inline src/Data/Hocker/Process.hs]: Hmm.. while unlikely to matter, you are aware that `T.pack` will munge your data if it is not good unicode right?

File: src/Data/Hocker/Process.hs (hunk 2 above)
Comment (lancelet) [inline src/Data/Hocker/Process.hs]: Gah - thanks, I should have known that by now. 😄

**Pull request merged**